### ISSUE TITLE
Rule suggestion: static-this

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [palantir/tslint#4428]
### Rule Suggestion

**Is your rule for a general problem or is it specific to your development style?**

General problem

**What does your suggested rule do?**

Ban the use of `this` in static methods.

Static `this` usage can be confusing for newcomers.
It can also become imprecise when used with extended classes when a static `this` of a parent class no longer specifically refers to the parent class.

**List several examples where your rule could be used**

Instead of:

```typescript
class Foo {
    private static value = true;
    static getValue() {
        return this.value;
    }
}
```

Prefer:

```typescript
class Foo {
    private static value = true;
    static getValue() {
        return Foo.value;
    }
}
```

See https://github.com/ajafff/tslint-consistent-codestyle/blob/master/docs/no-static-this.md

**Additional context**

Spinning off of #2040: moving some rules of tslint-consistent-codestyle to tslint core.
Note that as ajafff mentioned, copied code without substantial changes needs to retain the original license.

**Comment by mbelsky** [comment]
Hey,
I'm working on this rule's implementation. Could you please assign this issue to me?
Also if the 'static-this' name isn't final please let me know what do you think about name it 'no-static-this'?

**Comment by JoshuaKGoldberg** [comment]
@mbelsky there's no way to assign someone who's not in the Palantir org unfortunately. Posting here should be enough indicator to ask others not to tackle it soon, hopefully!

Re `no-static-this` vs `static-this`, is there a reason you'd prefer the name change? Prefixing rule names with `no-` has a couple downsides:
* It makes them less discoverable _(harder to scan through)_ in lists of rules
* It discourages adding options for the opposite style - in this case, if folks would prefer the `this` for brevity

**Comment by onehorsetown** [comment]
This breaks static functions that return an anonymous class, something that is typically used to construct higher-order components:

```
interface Bar {}

class Foo {
  public static higherOrderComponent(): any {
    return class implements Bar {
      public constructor() {
        console.log(this);   // "this" is valid here but tslint is complaining
      }
    };
  }
}
```

**Comment by mbelsky** [comment]
Hey,
Thanks for update. I'm going to bed now, so I'll try to fix that tomorrow if no one take it.

**Comment by mbelsky** [comment]
@JoshuaKGoldberg could you please create a new issue for this false positive case and assign to me?

**Comment by onehorsetown** [comment]
I opened a new issue:  #4579

I also noticed above that Joshua indicated that he could not assign something to you:

https://github.com/palantir/tslint/issues/4428#issuecomment-456125978

**Comment by JoshuaKGoldberg** [comment]
Oh yes, this issue was resolved by the original PR, and #4579 tracks the followup false positive case. Thanks folks!
